ckanext/qgov/common/anti_csrf.py

```diff
@@ -84,34 +84,45 @@ def get_cookie_token():
     return token
 
 def validate_token(token):
-    if not HMAC_PATTERN.match(token):
-        return false
-
     import time, hmac, hashlib, urllib
 
-    now = int(time.time())
-
-    parts = token.split('!', 1)
-    provided_hmac = unicode(parts[0])
-    message = parts[1]
+    token_values = read_token_values(token)
+    if not 'hash' in token_values:
+        return False
 
     secret_key = _get_secret_key()
-    expected_hmac = unicode(hmac.HMAC(secret_key, message, hashlib.sha512).hexdigest())
-    if not hmac.compare_digest(expected_hmac, provided_hmac):
+    expected_hmac = unicode(hmac.HMAC(secret_key, token_values['message'], hashlib.sha512).hexdigest())
+    if not hmac.compare_digest(expected_hmac, token_values['hash']):
         return False
 
-    message_parts = message.split('/', 2)
-    timestamp = int(message_parts[0])
-    username = message_parts[2]
-
+    now = int(time.time())
+    timestamp = token_values['timestamp']
     # allow tokens up to 30 minutes old
     if now < timestamp or now - timestamp > 60 * 30:
         return False
-    if username != urllib.quote(g.userobj.name):
+
+    if token_values['username'] != urllib.quote(g.userobj.name):
```

Comment:
why not flatten username to swap / with _ on the output and do the same when checking?

Reply:
Wouldn't that mean that a legitimate account containing a slash, such as "ab/cd", could be attacked by a malicious user registering "ab_cd" and obtaining a token that would be accepted for "ab/cd"?